arc

```diff
@@ -0,0 +1,163 @@
+#!/bin/bash
+
+# set -e
+
+usage () {
+  echo "Usage: $0 [OPTIONS] [check|build|run] <INPUT-FILE> [-- <MLIR-OPTS>]"
+  echo "Options:"
+  echo "  -h       Show this help message"
+  echo "  -o <DIR> Output directory (defaults to ./target/)"
+  exit 0
+}
+
+if [ "$DEBUG" ]; then
+  debug () {
+    echo "$@"
+  }
+else
+  debug () {
+    :
+  }
+fi
+
+while getopts ho: opt; do
+    case $opt in
+        h)
+            usage
+            ;;
+        o)
+            CRATE_DIR="$OPTARG"
+            ;;
+        \?)
+            echo "Invalid option: $OPTARG" >&2
+            usage
+            exit 1
+            ;;
+    esac
+done
+
+shift "$((OPTIND - 1))"
+
+debug "Parsed options"
+
+while [[ $# -gt 0 ]]; do
+    case $1 in
+        --)
+            shift
+            break
+            ;;
+        check)
+            MODE=check
+            INPUT_FILE="$1"
+            shift 3
+            break
+            ;;
+        build)
+            MODE=build
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        run)
+            MODE=run
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        *)
+            echo "Invalid subcommand: $1" >&2
+            usage
+            exit 1
+            ;;
+    esac
+done
+
+debug "Parsed subcommand"
+
+CRATE_NAME="$(basename $INPUT_FILE .arc)-crate"
+
+if [ -z "$INPUT_FILE" ]; then
+    usage
+fi
+
+if [ -z "$CRATE_DIR" ]; then
+    CRATE_DIR="$(dirname $INPUT_FILE)/$CRATE_NAME"
+fi
+
+ARC_DIR=$(realpath $(dirname "$(which "$0")")/)
+
+if [ -z "$ARC_LANG" ]; then
+    ARC_LANG="$ARC_DIR/arc-mlir/build/llvm-build/bin/arc-lang"
+    if [ ! -f "$ARC_LANG" ]; then
+        echo "Could not find arc-lang. Did you run './build'?"
+        exit 1
+    fi
+fi
+
+if [ -z "$ARC_MLIR" ]; then
+    ARC_MLIR="$ARC_DIR/arc-mlir/build/llvm-build/bin/arc-mlir"
+    if [ ! -f "$ARC_LANG" ]; then
+        echo "Could not find arc-mlir. Did you run './build'?"
+        exit 1
+    fi
+fi
+
+if [ -z "$ARC_RUNTIME_DIR" ]; then
+    ARC_RUNTIME_DIR="$ARC_DIR/arc-runtime/"
+fi
+
+INPUT_FILE="$(realpath $INPUT_FILE)"
+CRATE_DIR="$(realpath $CRATE_DIR)"
```

Comment:
Missing quoting.
Why are you using `realpath` here? You have already checked that the input file exists on a (perhaps convoluted) path? `CRATE_DIR` is also derived from `INPUT_FILE` so if you insist on `realpath`, why not do it once before defining `CRATE_DIR`?